GPUP CGContexts create unattributed Metal resources

kkinnunen-apple opened the pull request:
#### 869bc82ab5f697f13c23a833621ee77433b130fb
<pre>
GPUP CGContexts create unattributed Metal resources
<a href="https://bugs.webkit.org/show_bug.cgi?id=241509">https://bugs.webkit.org/show_bug.cgi?id=241509</a>
rdar://problem/94803367

Patch by Kimmo Kinnunen &lt;kkinnunen@apple.com &gt; on 2022-06-14
Reviewed by Dean Jackson.

Unattributed Metal related memory allocations grow GPUP footprint,
causing the process go over its memory limit.

Add a CoreGraphics SPI to mark that the CGContext resources belong to a
particular process.

Integrate this via IOSurface::setOwnershipIdentity(), as that is where
the current CGContext for IOSurfaces are instantiated and held.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/pal/spi/cg/CoreGraphicsSPI.h:
* Source/WebCore/PAL/pal/cg/CoreGraphicsSoftLink.cpp: Added.
* Source/WebCore/PAL/pal/cg/CoreGraphicsSoftLink.h: Added.
* Source/WebCore/PAL/pal/PlatformMac.cmake:
* Source/WebCore/platform/graphics/cocoa/IOSurface.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
(WebCore::IOSurface::ensurePlatformContext):
(WebCore::IOSurface::setOwnershipIdentity):

Canonical link: <a href="https://commits.webkit.org/251527@main">https://commits.webkit.org/251527@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@295522">https://svn.webkit.org/repository/webkit/trunk@295522</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

shallawa commented on Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
```diff
@@ -348,7 +348,10 @@ static IntSize computeMaximumSurfaceSize()
+#if HAVE(CGCONTEXT_SET_OWNER_IDENTITY)
+    if (m_resourceOwner)
+        CGContextSetOwnerIdentity(m_cgContext.get(), m_resourceOwner.taskIdToken());
+#endif
```
Is Metal used for IOSurface context based only? There are many cases where we create unaccelerated ImageBuffers (CGBitmapContext based ImageBuffers). Do we need to attribute the resources of these contexts to WebProcess as well?

Also should not GraphicsContext be the place we add this call to? So ensureGraphicsContext can look like this:

```
GraphicsContext& IOSurface::ensureGraphicsContext()
{
    if (m_graphicsContext)
        return *m_graphicsContext;

    m_graphicsContext = makeUnique<GraphicsContextCG>(ensurePlatformContext());
    m_graphicsContext->setIsAcceleratedContext(true);
    m_graphicsContext->setOwnerIdentity(m_resourceOwner);

    return *m_graphicsContext;
}   
```
And GraphicsContextCG::setOwnerIdentity() can have the `#if HAVE(CGCONTEXT_SET_OWNER_IDENTITY)` and the call to the SPI `CGContextSetOwnerIdentity()`.

kkinnunen-apple replied:
> Is Metal used for IOSurface context based only?

Yes, Metal is IOSurface only.

> There are many cases where we create unaccelerated ImageBuffers (CGBitmapContext based ImageBuffers). Do we need to attribute the resources of these contexts to WebProcess as well?

Later on, we may need to do that. Currently there is no mechanism to implement that, there is no implementation for it and we don't have data on what we would gain doing that.

> Also should not GraphicsContext be the place we add this call to? So ensureGraphicsContext can look like this:
>   m_graphicsContext->setOwnerIdentity(m_resourceOwner);

This would be the ideal, but given how the code is organised, this would look inconsistent and be incorrect.
From the code organisation perspective:
The CGContext is not owned by the GraphicsContextCG, and as such setting a persistent owner identity in GraphicsContextCG is inconsistent. Since the CGContext is not owned by GraphicsContextCG, it cannot alter persistent state for the CGontext. When CGContext is used for some other purpose, nothing guarantees that the owner identity is ok to set the way GraphicsContextCG mutates it. The owner identity cannot be unset.

From correctness perspective, we do not always have a GraphicsContextCG when doing operations on the CGContext. As such it's not ideal that we wouldn't use the attribution for the cases where we don't use  GraphicsContextCG.

The current code structure is such that the caller of GraphicsContextCG configures the underlying CGContext and then creates GraphicsContextCG. This patch adds to the configuration step.

shallawa replied:
I think you are right. 

CGContextSetOwnerIdentity() should be called from IOSurface::ensurePlatformContext() since  it is called from places other than IOSurface::ensureGraphicsContext(). So tying the memory attribution of CGContext to GraphicsContextCG is incorrect.

smfr commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -353,6 +353,13 @@
+#if (PLATFORM(MAC) && __MAC_OS_X_VERSION_MIN_REQUIRED >= 130000) \
+    || (((PLATFORM(IOS) && !PLATFORM(IOS_FAMILY_SIMULATOR)) || PLATFORM(MACCATALYST)) && __IPHONE_OS_VERSION_MIN_REQUIRED >= 160000) \
+    || (PLATFORM(APPLETV) && !PLATFORM(IOS_FAMILY_SIMULATOR) && __TV_OS_VERSION_MIN_REQUIRED >= 160000)
+// FIXME: Is this really unavailable in the iOS Simulator and watchOS?
```
Let's resolve the FIXME

Commit:
[RISC-V] Build fixes after 251455@main
https://bugs.webkit.org/show_bug.cgi?id=241591

Patch by Žan Doberšek <zdobersek@igalia.com> on 2022-06-14
Unreviewed, adding missing RISCV64 build guards alongside guards for other
64-bit platforms to get the build back up and running.

* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/wasm/WasmMemoryInformation.cpp:
(JSC::Wasm::PinnedRegisterInfo::get):

Canonical link: https://commits.webkit.org/251524@main
git-svn-id: https://svn.webkit.org/repository/webkit/trunk@295519 268f45cc-cd09-0410-ab3c-d52691b4dbfc

clopez committed:
run-benchmark script fails to find the lsof command on Linux
https://bugs.webkit.org/show_bug.cgi?id=241081

Reviewed by Dewei Zhu.

The lsof command is shipped on Linux typically on /usr/bin meanwhile on
Mac is shipped on /usr/sbin. Check if is on PATH before defaulting to the
Mac path.

* Tools/Scripts/webkitpy/benchmark_runner/http_server_driver/simple_http_server_driver.py:
  (SimpleHTTPServerDriver._find_http_server_port):

Canonical link: https://commits.webkit.org/251525@main
git-svn-id: https://svn.webkit.org/repository/webkit/trunk@295520 268f45cc-cd09-0410-ab3c-d52691b4dbfc

dpino committed:
[build.webkit.org] Add WebKitGTK and WPE bots for Ubuntu 22.04
https://bugs.webkit.org/show_bug.cgi?id=241335

Reviewed by Adrian Perez de Castro and Aakash Jain.

* Tools/CISupport/build-webkit-org/config.json: Add two new entries for
  building WebKitGTK and WPE on Ubuntu 20.04. The current Ubuntu LTS
  bots are now building WebKitGTK and WPE on Ubuntu 22.04.
* Tools/CISupport/build-webkit-org/factories_unittest.py: Add two new
  entries for testing WebKitGTK and WPE on Ubuntu 20.04.
* Tools/glib/dependencies/apt: Install package 'python-gi' only if available.
* Tools/gtk/dependencies/apt: Install packages only if available.

Canonical link: https://commits.webkit.org/251526@main
git-svn-id: https://svn.webkit.org/repository/webkit/trunk@295521 268f45cc-cd09-0410-ab3c-d52691b4dbfc

kkinnunen-apple committed:
GPUP CGContexts create unattributed Metal resources
https://bugs.webkit.org/show_bug.cgi?id=241509
rdar://problem/94803367

Patch by Kimmo Kinnunen <kkinnunen@apple.com> on 2022-06-14
Reviewed by Dean Jackson.

Unattributed Metal related memory allocations grow GPUP footprint,
causing the process go over its memory limit.

Add a CoreGraphics SPI to mark that the CGContext resources belong to a
particular process.

Integrate this via IOSurface::setOwnershipIdentity(), as that is where
the current CGContext for IOSurfaces are instantiated and held.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/pal/spi/cg/CoreGraphicsSPI.h:
* Source/WebCore/PAL/pal/cg/CoreGraphicsSoftLink.cpp: Added.
* Source/WebCore/PAL/pal/cg/CoreGraphicsSoftLink.h: Added.
* Source/WebCore/PAL/pal/PlatformMac.cmake:
* Source/WebCore/platform/graphics/cocoa/IOSurface.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
(WebCore::IOSurface::ensurePlatformContext):
(WebCore::IOSurface::setOwnershipIdentity):

Canonical link: https://commits.webkit.org/251527@main
git-svn-id: https://svn.webkit.org/repository/webkit/trunk@295522 268f45cc-cd09-0410-ab3c-d52691b4dbfc
